# -*- coding: utf-8 -*-
"""
åŠå…¬æ¡Œé£æ°´ gRPC æœåŠ¡å™¨
"""

import sys
import os
import logging
from concurrent import futures
import grpc

# æ·»åŠ é¡¹ç›®æ ¹ç›®å½•åˆ°è·¯å¾„
BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, BASE_DIR)
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

# å¯¼å…¥ç”Ÿæˆçš„ proto æ–‡ä»¶
try:
    from proto.generated import desk_fengshui_pb2, desk_fengshui_pb2_grpc
except ImportError:
    print("âš ï¸ protoæ–‡ä»¶æœªç”Ÿæˆï¼Œè¯·å…ˆè¿è¡Œç”Ÿæˆè„šæœ¬")
    # ä¸´æ—¶åˆ›å»ºå ä½ç¬¦
    desk_fengshui_pb2 = None
    desk_fengshui_pb2_grpc = None

from analyzer import DeskFengshuiAnalyzer

# é…ç½®æ—¥å¿—
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('logs/desk_fengshui.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)


class DeskFengshuiServicer:
    """åŠå…¬æ¡Œé£æ°´æœåŠ¡å®ç°"""
    
    def __init__(self):
        """åˆå§‹åŒ–æœåŠ¡"""
        self.analyzer = DeskFengshuiAnalyzer()
        logger.info("âœ… DeskFengshuiServicer åˆå§‹åŒ–æˆåŠŸ")
    
    def AnalyzeDesk(self, request, context):
        """åˆ†æåŠå…¬æ¡Œé£æ°´"""
        try:
            logger.info(f"æ”¶åˆ°åˆ†æè¯·æ±‚: use_bazi={request.use_bazi}")
            
            # è°ƒç”¨åˆ†æå™¨
            result = self.analyzer.analyze(
                image_bytes=request.image_data,
                solar_date=request.solar_date if request.solar_date else None,
                solar_time=request.solar_time if request.solar_time else None,
                gender=request.gender if request.gender else None,
                use_bazi=request.use_bazi
            )
            
            if not result.get('success'):
                return self._build_error_response(result.get('error', 'åˆ†æå¤±è´¥'))
            
            # æ„å»ºå“åº”
            response = self._build_success_response(result)
            
            logger.info(f"åˆ†ææˆåŠŸï¼Œè¯„åˆ†: {result.get('score', 0)}")
            
            return response
            
        except Exception as e:
            logger.error(f"åˆ†æè¯·æ±‚å¤„ç†å¤±è´¥: {e}", exc_info=True)
            return self._build_error_response(str(e))
    
    def HealthCheck(self, request, context):
        """å¥åº·æ£€æŸ¥"""
        if desk_fengshui_pb2:
            return desk_fengshui_pb2.HealthCheckResponse(status="healthy")
        else:
            # å ä½ç¬¦å®ç°
            class Response:
                status = "healthy"
            return Response()
    
    def _build_success_response(self, result: dict):
        """æ„å»ºæˆåŠŸå“åº”"""
        if not desk_fengshui_pb2:
            # å ä½ç¬¦å®ç°
            class Response:
                success = True
                score = 0
                summary = ""
                items = []
                adjustments = []
                additions = []
                removals = []
            return Response()
        
        # è½¬æ¢æ£€æµ‹åˆ°çš„ç‰©å“
        items = []
        for item in result.get('items', []):
            position = item.get('position', {})
            detected_item = desk_fengshui_pb2.DetectedItem(
                name=item.get('name', ''),
                label=item.get('label', ''),
                confidence=item.get('confidence', 0.0),
                bbox=item.get('bbox', []),
                relative_position=position.get('relative', ''),
                bagua_direction=position.get('bagua_direction', '')
            )
            items.append(detected_item)
        
        # è½¬æ¢è°ƒæ•´å»ºè®®
        adjustments = self._build_suggestions(result.get('adjustments', []))
        additions = self._build_suggestions(result.get('additions', []))
        removals = self._build_suggestions(result.get('removals', []))
        
        # æ„å»ºå“åº”
        response = desk_fengshui_pb2.DeskAnalysisResponse(
            success=True,
            items=items,
            adjustments=adjustments,
            additions=additions,
            removals=removals,
            score=result.get('score', 0),
            summary=result.get('summary', ''),
            xishen=result.get('xishen', ''),
            jishen=result.get('jishen', '')
        )
        
        return response
    
    def _build_suggestions(self, suggestions: list):
        """æ„å»ºå»ºè®®åˆ—è¡¨"""
        if not desk_fengshui_pb2:
            return []
        
        result = []
        for sugg in suggestions:
            suggestion = desk_fengshui_pb2.Suggestion(
                item=sugg.get('item', ''),
                item_label=sugg.get('item_label', ''),
                current_position=sugg.get('current_position', ''),
                ideal_position=sugg.get('ideal_position', sugg.get('position', '')),
                reason=sugg.get('reason', ''),
                priority=sugg.get('priority', ''),
                action=sugg.get('action', ''),
                element=sugg.get('element', '')
            )
            result.append(suggestion)
        
        return result
    
    def _build_error_response(self, error: str):
        """æ„å»ºé”™è¯¯å“åº”"""
        if not desk_fengshui_pb2:
            class Response:
                success = False
                error = ""
            resp = Response()
            resp.error = error
            return resp
        
        return desk_fengshui_pb2.DeskAnalysisResponse(
            success=False,
            error=error
        )


def serve(port: int = 9010):
    """å¯åŠ¨ gRPC æœåŠ¡å™¨ï¼ˆæ”¯æŒçƒ­æ›´æ–°ï¼‰"""
    if not desk_fengshui_pb2_grpc:
        logger.error("âŒ protoæ–‡ä»¶æœªç”Ÿæˆï¼Œæ— æ³•å¯åŠ¨æœåŠ¡")
        return
    
    try:
        from server.hot_reload.microservice_reloader import (
            create_hot_reload_server,
            register_microservice_reloader
        )
        
        server_options = [
            ('grpc.keepalive_time_ms', 300000),
            ('grpc.keepalive_timeout_ms', 20000),
        ]
        
        server, reloader = create_hot_reload_server(
            service_name="desk_fengshui",
            module_path="services.desk_fengshui.grpc_server",
            servicer_class_name="DeskFengshuiServicer",
            add_servicer_to_server_func=desk_fengshui_pb2_grpc.add_DeskFengshuiServiceServicer_to_server,
            port=port,
            server_options=server_options,
            max_workers=10,
            check_interval=30
        )
        
        register_microservice_reloader("desk_fengshui", reloader)
        reloader.start()
        
        # create_hot_reload_server å·²ç»ç»‘å®šäº†ç«¯å£ï¼Œä¸éœ€è¦å†æ¬¡ç»‘å®š
        logger.info(f"ğŸš€ åŠå…¬æ¡Œé£æ°´æœåŠ¡å¯åŠ¨åœ¨ç«¯å£ {port} (çƒ­æ›´æ–°å·²å¯ç”¨)")
        server.start()
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            logger.info("æœåŠ¡åœæ­¢")
            reloader.stop()
            server.stop(0)
            
    except ImportError:
        # é™çº§åˆ°ä¼ ç»Ÿæ¨¡å¼
        server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
        
        servicer = DeskFengshuiServicer()
        desk_fengshui_pb2_grpc.add_DeskFengshuiServiceServicer_to_server(servicer, server)
        
        server.add_insecure_port(f'[::]:{port}')
        
        logger.info(f"ğŸš€ åŠå…¬æ¡Œé£æ°´æœåŠ¡å¯åŠ¨åœ¨ç«¯å£ {port} (ä¼ ç»Ÿæ¨¡å¼)")
        server.start()
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            logger.info("æœåŠ¡åœæ­¢")
            server.stop(0)


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description='åŠå…¬æ¡Œé£æ°´ gRPC æœåŠ¡')
    parser.add_argument('--port', type=int, default=9010, help='æœåŠ¡ç«¯å£')
    
    args = parser.parse_args()
    
    serve(port=args.port)

